spec/models/game_spec.rb

```diff
@@ -1,11 +1,77 @@
+      expected = [
+        { col: 1, row: 1, color: 'white', type: 'Rook', found: false },
+        { col: 2, row: 1, color: 'white', type: 'Knight', found: false },
+        { col: 3, row: 1, color: 'white', type: 'Bishop', found: false },
+        { col: 4, row: 1, color: 'white', type: 'Queen', found: false },
+        { col: 5, row: 1, color: 'white', type: 'King', found: false },
+        { col: 6, row: 1, color: 'white', type: 'Bishop', found: false },
+        { col: 7, row: 1, color: 'white', type: 'Knight', found: false },
+        { col: 8, row: 1, color: 'white', type: 'Rook', found: false },
+
+        { col: 1, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 2, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 3, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 4, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 5, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 6, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 7, row: 2, color: 'white', type: 'Pawn', found: false },
+        { col: 8, row: 2, color: 'white', type: 'Pawn', found: false },
+
+        { col: 1, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 2, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 3, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 4, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 5, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 6, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 7, row: 7, color: 'black', type: 'Pawn', found: false },
+        { col: 8, row: 7, color: 'black', type: 'Pawn', found: false },
+
+        { col: 1, row: 8, color: 'black', type: 'Rook', found: false },
+        { col: 2, row: 8, color: 'black', type: 'Knight', found: false },
+        { col: 3, row: 8, color: 'black', type: 'Bishop', found: false },
+        { col: 4, row: 8, color: 'black', type: 'Queen', found: false },
+        { col: 5, row: 8, color: 'black', type: 'King', found: false },
+        { col: 6, row: 8, color: 'black', type: 'Bishop', found: false },
+        { col: 7, row: 8, color: 'black', type: 'Knight', found: false },
+        { col: 8, row: 8, color: 'black', type: 'Rook', found: false }
+      ]
+      # goes through each expected piece and find the first matching one in actual,
+      # then mark both found
+      expected.each do |e|
+        first_actual = actual.select do |a|
+          (a[:color] == e[:color]) && (a[:type] == e[:type]) && (a[:row] == e[:row]) && (a[:col] == e[:col])
+        end.first
+
+        e[:found] = true if first_actual
+      end
+      # expects that every expected piece was found.
+      missing = expected.select { |p| !(p[:found]) }
+      missing.each do |p|
+        puts "Missing piece #{p}"
+      end
+      expect(missing.count).to eq(0)
```

Comment:
Good test.

In future, another way of writing this that may end up a bit cleaner is.

Could use active record's `find_by` plus `exists?`

```Ruby
expect(game.pieces.black.find_by(type: 'Knight', x_coordinate: 0, y_coordinate: 0).exists?).to eq true
expect(game.pieces.black.find_by(type: 'King', x_coordinate: 0, y_coordinate: 0).exists?).to eq false
```

Other ways to use active record here would be

```Ruby
expect(game.pieces.black.where(type: 'Pawn').count).to eq 8
```